openforcefield/utils/toolkits.py

```diff
@@ -2948,6 +2949,161 @@ def compute_partial_charges_am1bcc(self, molecule):
+        supported_bond_order_models = list(bond_order_model_to_antechamber_keyword.keys())
+        bond_order_model = bond_order_model.lower()
+        if bond_order_model is None:
+            bond_order_model = 'am1-wiberg'
+        if bond_order_model not in supported_bond_order_models:
+            raise ValueError(f"Bond order model '{bond_order_model}' is not supported by AmberToolsToolkitWrapper. "
+                             f"Supported models are {supported_bond_order_models}")
+        ac_charge_keyword = bond_order_model_to_antechamber_keyword[bond_order_model]
+
+        from openforcefield.utils import temporary_directory, temporary_cd
+        with temporary_directory() as tmpdir:
+            with temporary_cd(tmpdir):
+                net_charge = molecule.total_charge
+                # Write out molecule in SDF format
+                self._rdkit_toolkit_wrapper.to_file(
+                    molecule, 'molecule.sdf', file_format='sdf')
+                # Prepare sqm.in file as if we were going to run charge calc
+                # TODO: Add error handling if antechamber chokes
+                subprocess.check_output([
+                    "antechamber", "-i", "molecule.sdf", "-fi", "sdf", "-o",
+                    "sqm.in", "-fo", "sqmcrt", "-pf", "yes", "-c", ac_charge_keyword,
+                    "-nc", str(net_charge)
+                ])
+                # Modify sqm.in to request bond order calculation
+                data = open('sqm.in').read()
+
+                # Original sqm.in file headerlooks like:
+
+                # Run semi-empirical minimization
+                #  &qmmm
+                #    qm_theory='AM1', grms_tol=0.0005,
+                #    scfconv=1.d-10, ndiis_attempts=700,   qmcharge=0,
+                #  /
+                # ... (atom coordinates in something like XYZ format) ...
+
+                # We need to add "printbondorders=1" to the list of keywords
+
+                # First, split the sqm.in text at the "/" mark at the end of the header
+                datasp = data.split("/")
+                # Insert the "printbondorders" directive in a new line and re-add the "/"
+                datasp.insert(1, 'printbondorders=1, \n /')
+                # Reassemble the file text
+                new_data = ''.join(datasp)
+                # Write the new file contents, overwriting the original file.
+                with open('sqm.in', 'w') as of:
+                    of.write(new_data)
+
+                # Run sqm to get bond orders
+                subprocess.check_output(["sqm", "-i", "sqm.in", "-o", "sqm.out", "-O"])
+
+                bond_orders = self._get_fractional_bond_orders_from_sqm_out('sqm.out')
+
+        # Note that sqm calculate WBOs for ALL PAIRS of atoms, not just those that have
+        # bonds defined in the original molecule. So here we iterate over the bonds in
+        # the original molecule and only nab the WBOs for those.
+        for bond in molecule.bonds:
```

Comment:
Is it possible this could go wrong and we pull out a bond order for the wrong pair of atoms due to file reordering somewhere in the workflow? Should we compare atom symbols as well to make sure the bond is correct?

Reply:
> Is it possible this could go wrong and we pull out a bond order for the wrong pair of atoms due to file reordering somewhere in the workflow

Yes. 

Right now, `AmberToolsToolkitWrapper.assign_fractional_bond_orders` and `compute_partial_charges` assume that the atom indexing is not modified by antechamber. I don't know if this is guaranteed everywhere, but nobody has reported a bug so far. The toolkit makes several assumptions like this, and beyond our "throw a lot of molecules through it, occasionally reversing the atom ordering, and see if it's alright" tests, I doubt even the AMBER manual says anything about this. 

So, this is something that occasionally keeps me up at night, but other than sifting through a bunch of fortran, I'm not sure that we can really prove this assumption. 

Long story short, "added in 43377fe "